daniela997/assignment-2/binary_tree.py

```
@@ -0,0 +1,121 @@
+        """
+        yield self
+        if self.left is not None:
+            for left in self.left.depth_first_traversal():
+                yield left
+        if self.right is not None:
+            for right in self.right.depth_first_traversal():
+                yield right
+
+    def breadth_first_traversal(self):
+        """Does breadth-first traversal of the subtree
+        to which the current node is a root. Uses a stack to keep track
+        of nodes to visit.
+        This is used as helper when adding new nodes to the tree.
+        :return: iterable containing BT nodes in breadth-first order.
+        """
+        to_visit = [self]
+        while to_visit:
+            current = to_visit.pop(0)
+            yield current
+            if current.left is not None:
+                to_visit.append(current.left)
+            if current.right is not None:
+                to_visit.append(current.right)
+
+    def append_node(self, new_node_key):
+        """Creates a new node with the given key and appends it
+        at the first "free" space using breadth_first_traversal
+        to find a "childless" parent. Assumes node keys are unique.
+        :param new_node_key: key value for new node to be added.
+        """
+        if new_node_key not in {node.key for node in self.breadth_first_traversal()}:
+            for parent in self.breadth_first_traversal():
+                if parent.left is None:
+                    parent.left = BinaryTree(new_node_key)
+                    break
+                elif parent.right is None:
+                    parent.right = BinaryTree(new_node_key)
+                    break
+
+    def find_node(self, search_key):
+        """Uses depth-first traversal to find a node with the given key.
+        :param search_key: find a node whose key is this value"""
+        for node in self.depth_first_traversal():
+            if node.key == search_key:
+                return node
+        return None
```

If the key is not found, then the user receives None. This means that None can not be a value anymore in this tree. Can you raise a KeyError exception instead? This is more in line with the list and the dict when accessing a value that does not exist.

---

I see a +1 to this comment, but the code stayed the same, so I am confused, can you clarify if you are going to heed the advice or have a rationale against it.

:)